include/deal.II/multigrid/mg_transfer_global_coarsening.templates.h

```diff
@@ -4102,6 +4107,135 @@ MGTransferMF<dim, Number>::initialize_internal_transfer(
 
 
 
+template <int dim, typename Number>
+std::pair<const DoFHandler<dim> *, unsigned int>
+MGTransferMF<dim, Number>::get_dof_handler_fine() const
+{
+  if (this->transfer.n_levels() <=
+      1) // single level: the information cannot be retreaved
+    return {nullptr, numbers::invalid_unsigned_int};
+
+  if (const auto t = dynamic_cast<
+        const MGTwoLevelTransfer<dim,
+                                 LinearAlgebra::distributed::Vector<Number>> *>(
+        this->transfer[this->transfer.max_level()].get()))
+    {
+      return {t->dof_handler_fine, t->mg_level_fine};
+    }
+  else if (const auto t = dynamic_cast<const MGTwoLevelTransferNonNested<
+             dim,
+             LinearAlgebra::distributed::Vector<Number>> *>(
+             this->transfer[this->transfer.max_level()].get()))
+    {
+      return {t->dof_handler_fine, t->mg_level_fine};
+    }
+  else
+    {
+      Assert(false, ExcNotImplemented());
+      return {nullptr, numbers::invalid_unsigned_int};
+    }
```

Comment:
I think the logic in this function is harder to read than necessary because you have two different return values but four cases. Can that we re-organized?

Reply:
How do you suggest? The problem is that the base class of `MGTwoLevelTransfer` and `MGTwoLevelTransferNonNested` does not have `dim` as template argument so that we need to cast to the two implementations.